Docker: add initial flow tests script

Galvien opened the pull request:
Add initial flow test script which will be used for testing that the easymesh flows are not broken.
For start, add a test to cover section 5.3.1 in WiFi certified EasyMesh test plan V1.0 - MCUT Supported services discovery test.

Signed-off-by: Gal Wiener <gal.wiener@intel.com>

Commit:
docker: add initial flow tests script

Add initial flow test script which will be used for testing that the
easymesh flows are not broken.
For start, add a test to cover section 5.3.1 in WiFi certified EasyMesh
test plan V1.0 - MCUT Supported services discovery test.

Signed-off-by: Gal Wiener <gal.wiener@intel.com>

tomereli commented:
Again Travis failed us with false positive.

arnout commented on tools/docker/tests/test_flows.sh:
```diff
@@ -0,0 +1,106 @@
+#find prplmesh folder
```
Missing copyright header (yes I know, a couple of other scripts are also still missing it).

tomereli replied:
Gal - please fix this and in a different PR add the copyright header to the other scripts which are missing it :)

arnout commented on tools/docker/tests/test_flows.sh:
```diff
@@ -0,0 +1,106 @@
+run() {
+    dbg "$*"
+    "$@" || exit $?
+}
```
Separate issue: we have a bunch of scripts that have the above boilerplate in various forms. Maybe it's time to refactor that...

tomereli replied:
First task for Coral! :)

arnout commented on tools/docker/tests/test_flows.sh:
```diff
@@ -0,0 +1,106 @@
+
+start_gw_repeater() {
+    dbg "delete running containers before starting test"
+    docker rm -f $(docker ps -a -q) >/dev/null 2>&1
```
I have other containers running on my system, so this is not OK...

Can't we just specify 'gateway' and 'repeater' here? Yes, it's possible that this was given a different name when launching the test_gw_repeater script, but we don't need to support that use case I think.

arnout commented on tools/docker/tests/test_flows.sh:
```diff
@@ -0,0 +1,106 @@
+
+send_topology_query() {
+    dbg "Sending toplogy query to al_mac $1"
+    docker exec -it gateway ${installdir}/bin/beerocks_cli -c "bml_trigger_topology_discovery $1" >/dev/null 2>&1
```
Do we need to redirect output here? It may be useful for debugging.

Perhaps we can define a global variable `redirect=">/dev/null 2>&1"` that gets set to empty if VERBOSE is on.

arnout commented on tools/docker/tests/test_flows.sh:
```diff
@@ -0,0 +1,106 @@
+
+check_log_for_query() {
+    dbg "Confirming topology query was received"
+    docker exec -it repeater cat tmp/$USER/beerocks/logs/beerocks_agent.log | grep -i -q "Topology Query"
```
No need for cat, just grep directly.

tomereli replied:
```suggestion
    docker exec -it repeater grep -i -q "Topology Query" tmp/$USER/beerocks/logs/beerocks_agent.log
```

tomereli replied:
Note - do not commit the suggestion as github offers - it adds a separate commit without DCO, so just overhead. Instead, change it in the file and add a fixup commit according to the CONTRIBUTING.md file.

arnout replied:
Actually, when you press "Commit suggestion" you can give the commit title, and there you can set the fixup message, i.e. "fixup! docker: add initial flow tests script".

And of course, you can anyway still move the fixup commit to the right place manually while doing git rebase -i.

arnout commented on tools/docker/tests/test_flows.sh:
```diff
@@ -0,0 +1,106 @@
+test_topology() {
+    status "test topology query"
+    send_topology_query $1
+    check_log_for_query
```
I don't think it makes much sense to have functions for send_topology_query and check_for_log_query - they will be used in only one place, and they're very simple.

Conversely, it might be useful to have functions for triggering a BML command:
```
send_bml_command() {
    docker exec -it gateway ${installdir}/bin/beerocks_cli -c "$*"
}
```
(note the `"$*"`, not the usual `"$@"`, because beerock_cli supports only one argument to -c).

The rest can just as well be inlined here directly I think.

(nitpick)

arnout commented on tools/docker/tests/test_flows.sh:
```diff
@@ -0,0 +1,106 @@
+        err "start GW+Repeater failed, abort"
+        exit 1
+    }
+    AL_MAC=$(docker exec -it gateway ${installdir}/bin/beerocks_cli -c bml_conn_map | grep IRE_BRIDGE | awk '{print $5}' | cut -d ',' -f 1)
```
Ouch. After my changes for the non-Intel agent, there won't be a IRE_BRIDGE in the topology any more...

I was thinking that we should use fixed MAC addresses instead of the random ones we have now. We can pass a four-byte base MAC address as an argument to the runner, and extend it with predictable sub-addresses for the various interfaces we create within the container.

But that's something for a later PR.

arnout replied:
I did this and the change was merged, so you can be sure that the gateway's AL_MAC is 00:11:22:33:00:00 and the repeater's is aa:bb:cc:dd:00:00.

Note that you need to refresh your docker images for that (or pull them from dockerhub).

Galvien replied:
Will fix in the next review iteration

arnout replied:
I hadn't noticed that this review comment wasn't taken into account yet. Anyway, there is no need to take into account changes that will be made by another PR. So feel free to leave this particular change for a future PR.

arnout commented on tools/docker/tests/test_flows.sh:
```diff
@@ -0,0 +1,106 @@
+    echo "      -h|--help - show this help menu"
+    echo "      -v|--verbose - verbosity on"
+    echo "      -a|--all - test all flows"
+    echo "      --test-topology - test the topology discovery flow"
```
I like this, having the option to test all or just one.

However, there probably will be many flows after some time. So maybe it is better to have positional arguments for the flows to run, so we don't have to pass it to getopt. And we should then of course use the name of the function `test_topology` so we don't have to match the string either.

tomereli replied:
I'm failing to understand the difference with having long options for running specific tests compared to having them as positional arguments - can you elaborate?

The use case we had in mind is to always use the --all option while developing, and see all of the flows which are not yet implemented as FAILURES.
Then for each new flow we will add the test and it will pass.

In travis, we will only use short options, so every time a new flow is ready, part of the PR is to update `.travis.yaml` to include that flow's test.

Finally - passing no positional arguments == `running test_gw_repeater --rm`

arnout replied:
Positional arguments -> call it like this:
```
test_flows.sh -v test-topology test-foo test-bar
```
i.e. without the `--`.

In code, it becomes completely different of course. You wouldn't have variables for the different tests, but just something like:
```
for test; do
    ${test}
done
```
(without an `in`, the for loop iterates over the positional arguments).

For handling the `-a` option, you'd do something like
```
set $all_tests
```
where `all_tests` is a variable containing the names of all tests. Obviously.

tomereli replied:
Oh I get it. makes the code much simpler.

arnout reviewed (changes requested):
Although all of my comments are just nitpicks, except for the copyright header, I'd still like the chance to review the second iteration so I don't approve yet.

tomereli reviewed (changes requested):
Now that full gw + repeater test is working in CI, we can add this to our CI as well.
Please also update .travis.yaml to run this script with the tests that should pass.
Every time a new test is added, .travis.yaml needs to also be updated. Make sure to add this to the CONTRIBUTING.md.

arnout commented on tools/docker/tests/test_flows.sh:
```diff
@@ -35,25 +42,86 @@ report() {
     dbg "delete running containers before starting test"
-    docker rm -f $(docker ps -a -q) >/dev/null 2>&1
+    docker rm -f gateway
+    docker rm -f repeater
```
I didn't know about the -f option - that's convenient! :+1:

arnout commented on tools/docker/tests/test_flows.sh:
```diff
@@ -35,25 +42,86 @@ report() {
+    docker rm -f repeater
     dbg "start gw and repeater"
-    ${scriptdir}/test_gw_repeater.sh >/dev/null 2>&1
+    ${scriptdir}/test_gw_repeater.sh -d 5 #>/dev/null 2>&1
```
Please don't leave commented-out code in the commit - or when you do, make sure that you document in the commit message why you have commented-out code.

arnout commented on tools/docker/tests/test_flows.sh:
```diff
@@ -35,25 +42,86 @@ report() {
-    check_log_for_query
+    send_bml_command "bml_trigger_topology_discovery $AL_MAC"
+    dbg "Confirming topology query was received"
+    docker exec -it repeater grep -i -q "Topology Query" tmp/$USER/beerocks/logs/beerocks_agent.log
```
Shouldn't this be `/tmp/...` ?

arnout commented on tools/docker/tests/test_flows.sh:
```diff
@@ -85,21 +166,20 @@ main() {
+    done
 }
 
+ALL_TESTS="topology initial_ap_config ap_config_renew ap_config_bss_tear_down channel_selection ap_capability_info_reporting client_capability_info_reporting link_metric_collection steering_mandate_opportunity client_steering client_association backhaul_optimization layer_data_payload_trigger layer_data_payload"
```
Two minor nits here:

- I find it more logical to put the defaults that may be overridden by command-line options here, but to put 'global const' variables in the beginning of the script (i.e. before defining functions, i.e. the place where you'd define global variables in C code). But this is just personal preference.
- The long line should be split. Since it's between quotes and since it will be whitespace-expanded where it is used, you can just split it anywhere, so:

```
ALL_TESTS="topology initial_ap_config ap_config_renew ap_config_bss_tear_down
    channel_selection ap_capability_info_reporting client_capability_info_reporting
    ..."
```

(I didn't count the characters here, but we wrap at 100 columns, cfr. .clang-format.)

arnout reviewed (approved):
I have a couple of minor nits. Please rebase and fix up while rebasing (or, if you disagree with my comments, don't fix it up). I'm approving already so it can be merged as soon as the checks succeed.

Commit:
travis: replace test_gw_repeater.sh with test_flows.sh

Now that we have a full test to verify all flows, use that in travis CI.
Have travis test perform the topology test which is the only test
currently implemented. From now on, every new flow will include adding a
test to test_flows.sh and update travis accordingly.

Also update CONTRIBUTING.md accordingly.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>
